# Containers-only Incus for tachyon (NanoPi R3S LTS, headless router).
# The stock module wires the full VM stack into the daemon closure; this
# host runs containers only. Prize: ~850-950 MiB off the system closure.
# Conscious trade-offs (each revertible by restoring one line):
# no `incus launch --vm`, no OCI import (skopeo/umoci), no stateful
# snapshots (criu), no LVM/ext4/XFS pool tools, no ACME via lego, no
# wifi (iw), no `incus documentation` (doc), no VM console (spice-gtk).
diff --git a/nixos/modules/virtualisation/incus.nix b/nixos/modules/virtualisation/incus.nix
--- a/nixos/modules/virtualisation/incus.nix
+++ b/nixos/modules/virtualisation/incus.nix
@@ -5,6 +5,9 @@
   ...
 }:
 
+# NOTE (patched for tachyon): containers-only policy, see
+# patches/incus-containers-only.patch. Every VM-only reference below
+# (qemu, OVMF, swtpm, virtiofsd, seabios, spice-gtk, docs) is stripped.
 let
   cfg = config.virtualisation.incus;
 
@@ -20,56 +23,32 @@
       cfg.package
       "/run/wrappers"
 
-      # some qemu helpers not in bin
-      (linkFarm "incus-qemu-libexec" [
-        {
-          name = "bin";
-          path = "${qemu_kvm}/libexec";
-        }
-      ])
-
       acl
       attr
       bash
       btrfs-progs
       bzip2
-      cdrkit
       coreutils
-      criu
       dnsmasq
-      e2fsprogs
       findutils
       gawk
       getent
       gnugrep
       gnused
       gnutar
-      gptfdisk
       gzip
       iproute2
       iptables
-      iw
       kmod
-      lego
-      libxfs
-      lvm2
       lxcfs
       lz4
       nftables
-      qemu-utils
-      qemu_kvm
       rsync
-      skopeo
       squashfs-tools-ng
       squashfsTools
       sshfs
-      swtpm
       systemd
-      thin-provisioning-tools
-      umoci
       util-linux
-      virtiofsd
-      xdelta
       xz
       zstd
     ]
@@ -95,53 +74,8 @@
       libnvidia-container
     ];
 
-  # https://github.com/lxc/incus/blob/cff35a29ee3d7a2af1f937cbb6cf23776941854b/internal/server/instance/drivers/driver_qemu.go#L123
-  OVMF2MB = pkgs.OVMF.override {
-    secureBoot = true;
-    fdSize2MB = true;
-  };
-  ovmf-prefix = if pkgs.stdenv.hostPlatform.isAarch64 then "AAVMF" else "OVMF";
-  ovmf = pkgs.linkFarm "incus-ovmf" (
-    [
-      # 2MB must remain the default or existing VMs will fail to boot. New VMs will prefer 4MB
-      {
-        name = "OVMF_CODE.fd";
-        path = "${OVMF2MB.fd}/FV/${ovmf-prefix}_CODE.fd";
-      }
-      {
-        name = "OVMF_VARS.fd";
-        path = "${OVMF2MB.fd}/FV/${ovmf-prefix}_VARS.fd";
-      }
-      {
-        name = "OVMF_VARS.ms.fd";
-        path = "${OVMF2MB.fd}/FV/${ovmf-prefix}_VARS.fd";
-      }
-
-      {
-        name = "OVMF_CODE.4MB.fd";
-        path = "${pkgs.OVMFFull.fd}/FV/${ovmf-prefix}_CODE.fd";
-      }
-      {
-        name = "OVMF_VARS.4MB.fd";
-        path = "${pkgs.OVMFFull.fd}/FV/${ovmf-prefix}_VARS.fd";
-      }
-      {
-        name = "OVMF_VARS.4MB.ms.fd";
-        path = "${pkgs.OVMFFull.fd}/FV/${ovmf-prefix}_VARS.fd";
-      }
-    ]
-    ++ lib.optionals pkgs.stdenv.hostPlatform.isx86_64 [
-      {
-        name = "seabios.bin";
-        path = "${pkgs.seabios-qemu}/share/seabios/bios.bin";
-      }
-    ]
-  );
-
   environment = lib.mkMerge [
     {
-      INCUS_DOCUMENTATION = "${cfg.package.doc}/html";
-      INCUS_EDK2_PATH = ovmf;
       INCUS_LXC_HOOK = "${cfg.lxcPackage}/share/lxc/hooks";
       INCUS_LXC_TEMPLATE_CONFIG = "${pkgs.lxcfs}/share/lxc/config";
       INCUS_USBIDS_PATH = "${pkgs.hwdata}/share/hwdata/usb.ids";
@@ -343,9 +277,6 @@
 
     environment.systemPackages = [
       cfg.clientPackage
-
-      # gui console support
-      pkgs.spice-gtk
     ];
 
     # Note: the following options are also declared in virtualisation.lxc, but
@@ -395,14 +326,8 @@
         r ${pkgs.stdenv.cc.libc}/lib/gconv/gconv-modules.d/,
         r ${pkgs.stdenv.cc.libc}/lib/gconv/gconv-modules.d/gconv-modules-extra.conf,
 
-        # Support use of VM instance
-        mrix ${pkgs.qemu_kvm}/bin/*,
-        k ${OVMF2MB.fd}/FV/*.fd,
-        k ${pkgs.OVMFFull.fd}/FV/*.fd,
       ''
-      + lib.optionalString pkgs.stdenv.hostPlatform.isx86_64 ''
-        k ${pkgs.seabios-qemu}/share/seabios/bios.bin,
-      '';
+      ;
     };
 
     security.acme.certs = lib.mkIf (cfg.useACMEHost != null) {
